### PR TITLE
Triton flash attention

### PR DESCRIPTION
**Pull request by kelvin-zou**
This PR contains 2 changes:
1. Fix the NaN issue on flash attention
2. Add a segment ID (2d-bias) support in flash attention.

Also contains updated tests, and refactoring the existing code.

### CODE REVIEW

**Comment by samos123** [comment]
How would I enable it in fuji 7b? Could you add a config for fuji 7b on GPU that uses flash attention?

File: pyproject.toml [hunk 1]
```diff
@@ -109,6 +109,10 @@ dataflow = [
     "google-apitools",  # for beam pipeline
     "orjson==3.9.10",
 ]
+# Triton kernel dependency.
+triton = [
```
Comment (samos123) [inline pyproject.toml:113]: maybe rename `triton` to `gpu` if this is a gpu specific dependency? Not sure if there will be other gpu only dependencies

**Review by markblee — APPROVED** [review]
Thanks, can we internally link the tb with the comparison between flash/no-flash?

**Comment by markblee** [comment]
> How would I enable it in fuji 7b? Could you add a config for fuji 7b on GPU that uses flash attention?

The [FlashAttention](https://github.com/apple/axlearn/blob/ef4c3820e92b0d4d83db23554eed2fdacf9d5f72/axlearn/common/flash_attention/layer.py#L37) layer should be (mostly) a drop-in replacement for `MultiheadAttention`/`GroupedQueryAttention` layers, with caveats mentioned in the class docstring. 

You can follow a similar idea as [set_norm_recursively](https://github.com/apple/axlearn/blob/ef4c3820e92b0d4d83db23554eed2fdacf9d5f72/axlearn/common/layers.py#L1593) to replace instances of `MultiheadAttention.Config` with your flash attention config.

**Comment by kelvin-zou** [comment]
> Thanks, can we internally link the tb with the comparison between flash/no-flash?

Done, please check internal PR.

**Review by markblee — APPROVED** [review]

**Pull request merged**